### PR TITLE
Reorganize project for updated tooling

### PR DESCRIPTION
**Pull request by grovduck**
This PR updates the tooling on this project to take advantage of newer tooling (like `hatch` and `pre-commit`) as used in [`sknnr`](https://github.com/lemma-osu/sknnr).  Pre-commits include `ruff`, `ruff-format`, and `mypy`.  Older setup specified in files like `setup.py` and `requirements.txt` have now been replaced by `pyproject.toml`.  Additionally, add Github actions to test with `[3.9, 3.10, 3.11, 3.12]`.

A checklist of items accomplished with this PR:
- [x] Remove requirements and setup files
- [x] Add `.pre-commit-config.yaml` file
- [x] Update `pyproject.toml` file
- [x] Move all package code under a `src` folder
- [x] Lint existing code
- [x] Convert `README.rst` to `README.md`
- [x] Add `CONTRIBUTING.md` to detail tooling setup
- [x] Replace GDAL utilities
- [x] Add Github actions

### CODE REVIEW

**Comment by grovduck** [comment]
Because `gdal` does not build easily on Windows, we need to replace all functionality that relies on this package and remove `gdal` from the dependencies.

**Comment by grovduck** [comment]
@aazuspan, I'm hoping you might be able to give this PR a very cursory look, specifically with the build elements.  The code is still a total mess, but I'm hoping the scaffolding of `hatch`, `pre-commit`, and Github actions will make changes going forward easier to handle.  Most of the changes here are linting changes that I've corrected just to pass linting, so please don't bother looking too closely at that unless you want to.

**Review by aazuspan — APPROVED** [review]
I took a *very* shallow look and just pointed out a couple little nits, but I was able to get everything set up on my machine and run tests/pre-commits/CLI commands, so I'd say LGTM!

File: pyproject.toml [hunk 1]
```diff
@@ -1,3 +1,102 @@
...
+    "rpy2",
+    "scikit-learn",
+    "scipy",
+]
+
+[project.scripts]
+pynnmap = "pynnmap.cli.main:main_group"
+
+[project.entry-points.pynnmap]
+build_attribute_raster = "pynnmap.cli.build_attribute_raster:cli_main"
```
[the remark this answers is not in the document]
Comment (grovduck) [reply pyproject.toml]: Good catch.  For now, I've renamed all functions to `main` within the modules, so they are at least consistent.

**Comment by grovduck** [comment]
> I took a _very_ shallow look and just pointed out a couple little nits, but I was able to get everything set up on my machine and run tests/pre-commits/CLI commands, so I'd say LGTM!

Thanks @aazuspan, exactly what I was hoping for!

**Pull request merged**